src/models/metrics.py

```diff
@@ -0,0 +1,137 @@
+"""Functions for calculating metrics from a GeoGraph."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Optional
+
+import networkx as nx
+import numpy as np
+
+if TYPE_CHECKING:
+    from src.models import geograph
+
+
+# define a metric dataclass with < <= => > == comparisons that work as you would
+# expect intuitively
+@dataclass()
+class Metric:
+    """Class to represent a metric for a GeoGraph, with associated metadata."""
+
+    value: Any  # No good Numpy type hints
+    name: str
+    description: str
+    variant: Optional[str]
+    unit: Optional[str] = None
+
+    def __eq__(self, o: object) -> bool:
+        if not isinstance(o, Metric):
+            return False
+        return self.value == o.value
+
+    def __lt__(self, o: object) -> bool:
+        if not isinstance(o, Metric):
+            return False
+        return self.value < o.value
+
+    def __le__(self, o: object) -> bool:
+        if not isinstance(o, Metric):
+            return False
+        return self.value <= o.value
+
+    def __gt__(self, o: object) -> bool:
+        if not isinstance(o, Metric):
+            return False
+        return self.value > o.value
+
+    def __ge__(self, o: object) -> bool:
+        if not isinstance(o, Metric):
+            return False
+        return self.value >= o.value
+
+
+def _num_components(geo_graph: geograph.GeoGraph) -> Metric:
+    return Metric(
+        value=nx.number_connected_components(geo_graph.graph),
+        name="num_components",
+        description="The number of connected components in the graph.",
+        variant="component",
+    )
+
+
+def _avg_patch_area(geo_graph: geograph.GeoGraph) -> Metric:
+    return Metric(
+        value=np.mean(geo_graph.df.area.values),
+        name="avg_patch_area",
+        description="The average area of the patches in the graph.",
+        variant="conventional",
+    )
+
+
+def _total_area(geo_graph: geograph.GeoGraph) -> Metric:
+    return Metric(
+        value=np.sum(geo_graph.df.area.values),
+        name="total_area",
+        description="The total area of all the patches in the graph.",
+        variant="conventional",
+    )
+
+
+def _avg_component_area(geo_graph: geograph.GeoGraph) -> Metric:
+    comp_geograph = geo_graph.components
+    if not comp_geograph.has_df:
```

Comment:
Question for usage: does a `GeoGraph` instance by default have this attribute, with a `ComponentGeoGraph` with df? Or how can I make sure that this metric can be computed for a GeoGraph?

Reply:
By default in the `GeoGraph` init we call `self.components = self.get_graph_components(calc_polygons=False)`, which means the resulting `ComponentsGeoGraph` does not have a df so this metric will not work there. You can calculate the df and store it in `components` with `geograph.components = geograph.get_graph_components(calc_polygons=True)`. However, by default the `HabitatGeoGraph` does get the components along with the component df in its init function, so this metric will work fine for those.